geniusyield-market-maker/src/GeniusYield/MarketMaker/Prices.hs

```diff
@@ -65,6 +142,285 @@ data MMToken = MMToken
+    isRelevantPairInfo ∷ MMTokenPair → DexPairInfo → Maybe (DexPairInfo, Bool)
+    isRelevantPairInfo mmtp dpi@DexPairInfo {..} =
+      ( (dpi, False)
+          <$ findMatchingMMTP mmtp
+            (dexPairInfoCoinAAssetName, dexPairInfoCoinAPolicy)
+            (dexPairInfoCoinBAssetName, dexPairInfoCoinBPolicy)
+      )
+        <|> ( (dpi, True)
+                <$ findMatchingMMTP mmtp
+                  (dexPairInfoCoinBAssetName, dexPairInfoCoinBPolicy)
+                  (dexPairInfoCoinAAssetName, dexPairInfoCoinAPolicy)
+            )
+
+    findMatchingMMTP ∷ MMTokenPair → (TokenName, PolicyId) → (TokenName, PolicyId) → Maybe MMTokenPair
+    findMatchingMMTP mmtp tokenA tokenB = fromRight Nothing $ do
+      assetClassA ← assetClassFromMaestro tokenA
+      assetClassB ← assetClassFromMaestro tokenB
+      Right $ if assetClassA == mmtAc (mmtpCurrency mmtp) && assetClassB == mmtAc (mmtpCommodity mmtp) then Just mmtp else Nothing
+
+    functionLocationIdent = "getMaestroPrice"
+
+buildGetQuota PriceCommonCfg {..} (TaptoolsPPC TaptoolsConfig {..}) = GetQuota $ \mmtp → case pccNetworkId of
+  GYMainnet → do
+    commodity ∷ MMToken ← 
+      case ttcPairOverride of
+        Just ttpo → pure MMToken {mmtAc = ttpoAsset ttpo, mmtPrecision = ttpoPrecision ttpo}
+        Nothing
+          | mmtpCurrency mmtp  == mmtLovelace → pure . mmtpCommodity $ mmtp
+          | mmtpCommodity mmtp == mmtLovelace → pure . mmtpCurrency $ mmtp
+          | otherwise                         → throwIO $ userError "Trading commodity pairs (non-ADA) not yet supported."
+
+    manager' ← taptoolsManager
+    let env = mkClientEnv manager' taptoolsBaseUrl
+
+    resolution ← maybe (fromCommonResolution pccCommonResolution) pure ttcResolutionOverride
+
+    case commodity of
+      MMToken { mmtAc = GYLovelace }                              → do
+        return . Right $ Price (1 % 1)
+
+      MMToken { mmtAc = GYToken cs tn, mmtPrecision = precision } → do
+        let cs'       = withoutQuotes . show $ cs
+            tn'       = withoutQuotes . show . tokenNameToHex $ tn
+            unit      = cs' ++ tn'
+        
+        ohlcvInfo ← priceFromTaptools ttcApiKey unit resolution 1 env
+
+        case ohlcvInfo of
+          Left e   → return . Left $ SourceUnavailable (displayException e) "Taptools"
+          Right [] → return . Left $ SourceUnavailable "Empty OHLCV." "Taptools"
+          Right (ttOHLCV : _) → do
+            let price         = close ttOHLCV
+                precisionDiff = 10 ** fromIntegral (mmtPrecision mmtLovelace - precision)
+                adjustedPrice = price * precisionDiff
+            return . Right . Price . toRational $ adjustedPrice
+        where
+          withoutQuotes ∷ String → String
+          withoutQuotes s = case s of
+            ('"':xs) | not (null xs) && last xs == '"' → init xs
+            _                                          → s
+
+  _         → throwIO $ userError "Only Mainnet is supported."
```

Comment:
Preferably an exception type for TapTools.

Reply:
Perhaps could be part of earlier checks like `buildPP`?

Reply:
Done:  now check is part of `buildPP`.